File: frontend/src/common/constants/propertyEvaluation.ts
/**
 * Property Evaluation Calculator constants
 * Default values for investment property analysis
 */

export const EVALUATION_DEFAULTS = {
  // Transaction costs (percentages)
  BROKER_FEE_PERCENT: 3.57,
  NOTARY_FEE_PERCENT: 1.5,
  LAND_REGISTRY_FEE_PERCENT: 0.5,

  // Depreciation and building
  DEPRECIATION_RATE_PERCENT: 2.0,
  BUILDING_SHARE_PERCENT: 70,

  // Forecast assumptions (annual percentages)
  COST_INCREASE_PERCENT: 2.0,
  RENT_INCREASE_PERCENT: 2.0,
  VALUE_INCREASE_PERCENT: 1.5,

  // Tax and equity (under Rent, Taxes, Forecast section)
  EQUITY_INTEREST_PERCENT: 5.0,
  MARGINAL_TAX_RATE_PERCENT: 42,

  // Financing defaults
  LOAN_PERCENT: 100,
  INTEREST_RATE_PERCENT: 4.0,
  REPAYMENT_RATE_PERCENT: 2.0,

  // Operating costs defaults (EUR/month absolute - user enters from Abrechnung)
  HAUSGELD_ALLOCABLE: 0,
  PROPERTY_TAX_MONTHLY: 0,
  HAUSGELD_NON_ALLOCABLE: 0,
  RESERVES_PORTION: 0,

  // Rent defaults
  RENT_PER_SQM: 12.0,
  PARKING_RENT: 50,
} as const

export const SECTION_COLORS = {
  propertyInfo:
    "bg-yellow-100 dark:bg-yellow-900/30 text-yellow-800 dark:text-yellow-400",
  rent: "bg-green-100 dark:bg-green-900/30 text-green-800 dark:text-green-400",
  operatingCosts:
    "bg-orange-100 dark:bg-orange-900/30 text-orange-800 dark:text-orange-400",
  financing:
    "bg-purple-100 dark:bg-purple-900/30 text-purple-800 dark:text-purple-400",
  evaluation: "bg-red-100 dark:bg-red-900/30 text-red-800 dark:text-red-400",
} as const

File: backend/app/schemas/property_evaluation.py
"""Property evaluation calculator request/response schemas."""

import dataclasses
import typing
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, create_model

from app.services.property_evaluation_calculator import (
    AnnualCashflowRow,
    EvaluationResult,
)

# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _pydantic_from_dataclass(
    dc_cls: type,
    model_name: str,
    *,
    exclude_fields: set[str] | None = None,
    overrides: dict[str, tuple] | None = None,
) -> type[BaseModel]:
    """Create a Pydantic model mirroring a dataclass's field definitions.

    Uses ``typing.get_type_hints()`` to resolve forward-ref annotations
    (from ``__future__ import annotations`` in the calculator module).
    """
    exclude = exclude_fields or set()
    hints = typing.get_type_hints(dc_cls)
    field_defs: dict[str, tuple] = {}

    for f in dataclasses.fields(dc_cls):
        if f.name in exclude:
            continue
        default = (
            f.default
            if f.default is not dataclasses.MISSING
            else (
                f.default_factory()
                if f.default_factory is not dataclasses.MISSING
                else ...
            )
        )
        field_defs[f.name] = (hints[f.name], default)

    if overrides:
        field_defs.update(overrides)

    return create_model(
        model_name,
        __config__=ConfigDict(from_attributes=True),
        **field_defs,
    )


# ---------------------------------------------------------------------------
# /calculate endpoint schemas
# ---------------------------------------------------------------------------


class PropertyEvaluationCalculateRequest(BaseModel):
    """Flat request model for the /calculate endpoint.

    All percent-scale values (e.g. broker_fee_percent=3.57 means 3.57%).
    Every field has a default so the frontend only sends what it has.
    """

    # Property details
    address: str = ""
    square_meters: float = 0.0
    purchase_price: float = 0.0
    rent_per_m2: float = 0.0
    parking_space_rent: float = 0.0

    # Transaction cost rates (percent scale, e.g. 3.57)
    broker_fee_percent: float = 3.57
    notary_fee_percent: float = 1.5
    land_registry_fee_percent: float = 0.5
    property_transfer_tax_percent: float = 5.0

    # Monthly management costs (EUR/month)
    base_allocable_costs: float = 0.0
    property_tax_monthly: float = 0.0
    base_non_allocable_costs: float = 0.0
    reserves_monthly: float = 0.0

    # Depreciation (percent scale)
    building_share_percent: float = 70.0
    afa_rate_percent: float = 2.0

    # Financing (percent scale)
    loan_percent: float = 100.0
    interest_rate_percent: float = 4.0
    initial_repayment_rate_percent: float = 2.0

    # Tax (percent scale)
    personal_taxable_income: float = 0.0
    marginal_tax_rate_percent: float = 42.0

    # Growth assumptions (percent scale)
    cost_increase_percent: float = 2.0
    rent_increase_percent: float = 2.0
    value_increase_percent: float = 2.0
    equity_interest_percent: float = 5.0

    # Renovation
    renovation_year: int = 0
    renovation_cost: float = 0.0

    # Analysis configuration
    start_year: int = Field(default_factory=lambda: datetime.now().year)
    analysis_years: int = 11


AnnualCashflowRowResponse = _pydantic_from_dataclass(
    AnnualCashflowRow,
    "AnnualCashflowRowResponse",
)

PropertyEvaluationCalculateResponse = _pydantic_from_dataclass(
    EvaluationResult,
    "PropertyEvaluationCalculateResponse",
    exclude_fields={"annual_rows"},
    overrides={"annual_rows": (list[AnnualCashflowRowResponse], [])},
)


# ---------------------------------------------------------------------------
# CRUD schemas
# ---------------------------------------------------------------------------


class PropertyEvaluationCreate(BaseModel):
    name: str | None = Field(None, max_length=255)
    journey_step_id: uuid.UUID | None = None
    inputs: dict = Field(..., description="Full PropertyEvaluationState from frontend")


class PropertyEvaluationResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str | None = None
    share_id: str | None = None
    journey_step_id: uuid.UUID | None = None
    purchase_price: float
    square_meters: float
    state_code: str | None = None
    cashflow_after_tax: float
    gross_rental_yield: float
    return_on_equity: float
    is_positive_cashflow: bool
    inputs: dict
    results: dict
    created_at: datetime


class PropertyEvaluationSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str | None = None
    share_id: str | None = None
    purchase_price: float
    cashflow_after_tax: float
    gross_rental_yield: float
    is_positive_cashflow: bool
    created_at: datetime


class PropertyEvaluationListResponse(BaseModel):
    data: list[PropertyEvaluationSummary]
    count: int
